packages/proxy/src/handlers/proxyRequests.ts

```diff
@@ -0,0 +1,52 @@
+import { URL } from 'node:url';
+import { DiscordAPIError, RateLimitError, RequestMethod, REST, RouteLike } from '@discordjs/rest';
+import {
+	populateGeneralErrorResponse,
+	populateOkResponse,
+	populateRatelimitErrorResponse,
+} from '../util/responseHelpers';
+import type { RequestHandler } from '../util/util';
+
+/**
+ * Creates an HTTP handler used to forward requests to Discord
+ * @param rest REST instance to use for the requests
+ */
+export function proxyRequests(rest: REST): RequestHandler {
+	return async (req, res) => {
+		const { method, url } = req;
+
+		if (!method || !url) {
+			throw new TypeError(
+				'Invalid request. Missing method and/or url, implying that this is not a Server IncomingMesage',
+			);
+		}
+
+		// The 2nd parameter is here so the URL constructor doesn't complain about an "invalid url" when the origin is missing
+		// we don't actually care about the origin and the value passed is irrelevant
+		const fullRoute = new URL(url, 'http://noop').pathname.replace(/^\/api(\/v\d+)?/, '') as RouteLike;
+
+		try {
+			const discordResponse = await rest.raw({
+				body: req,
+				fullRoute,
+				// This type cast is technically incorrect, but we want Discord to throw Method Not Allowed for us
+				method: method as RequestMethod,
+				passThroughBody: true,
+			});
+
+			await populateOkResponse(res, discordResponse);
+		} catch (error) {
+			if (error instanceof DiscordAPIError) {
+				populateGeneralErrorResponse(res, error);
+			} else if (error instanceof RateLimitError) {
+				populateRatelimitErrorResponse(res, error);
+			} else {
+				// Unclear if there's better course of action here. Any web framework allows to pass in an error handler for something like this
+				// at which point the user could dictate what to do with the error - otherwise we could just 500
+				throw error;
+			}
```

Comment:
Other errors thrown here are `AbortError` (thrown by [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)) and [`HTTPError`](https://github.com/discordjs/discord.js/blob/main/packages/rest/src/lib/errors/HTTPError.ts).

What's the behaviour in the proxy when either of those two happen?

Reply:
For AbortError I'm unsure what the behavior should be, now that you mention it.

Right now the behavior would be essentially silent failure on the proxy end which would cause god knows what on the client side.

I think the fix is to make the proxy NOT abort (i.e. make the timeout fairly high) so the client can do it.

So say (just throwing numbers out, I don't know what the defaults are right now), the proxy aborts after 2 minutes, meaning the only supported behavior is to have clients abort in less than that.

No other good way I can think of to deal with this - other than intentionally letting the request hang when an AbortError is thrown - but that seems equally scuffed.


As per HTTPError, I missed that. When is it thrown? 5xx after multiple retries? I thought that'd fall under DiscordAPIError initially. Will cover this.

Reply:
AbortError may return 408 (Request Timeout), we will need to handle this in /rest. Likely also make a TimedOutError.

> The server timed out waiting for the request. According to HTTP specifications: "The client did not produce a request within the time that the server was prepared to wait. The client MAY repeat the request without modifications at any later time."

Although I think 504 (Gateway Timeout) would probably be better, but it mixes up with Discord's errors, which can return this.

> The server was acting as a gateway or proxy and did not receive a timely response from the upstream server.

There's also nginx's unofficial code, 444 (No Response):

> Used internally to instruct the server to return no information to the client and close the connection immediately.

And Cloudflare's 524 (A Timeout Occurred), but might mix up with Discord too:

> Cloudflare was able to complete a TCP connection to the origin server, but did not receive a timely HTTP response.

We could also use an error Discord uses and make /rest differentiate between the two via headers, but I don't think it's really a good idea.

---

HTTPError is thrown on 5XX, it's used specifically because those generally don't give us a JSON body.

Also, I was thinking of also using error.cause, which is available since Node.js v16.10.0, in case we want to expose the original error as well, or we could just do it, since it's backwards compatible.